src/atomics/relaxed.md

```diff
@@ -0,0 +1,423 @@
+
+```rust
+# use std::sync::atomic::{self, AtomicU64};
+static COUNTER: AtomicU64 = AtomicU64::new(0);
+pub fn get_id() -> u64 {
+    COUNTER.fetch_add(1, atomic::Ordering::Relaxed)
+}
+```
+
+And then, no matter how many threads there are, that race condition from earlier
+can never occur. Executions will have to look more like this:
+
+```text
+  Thread 1     COUNTER     Thread 2
+╭───────────╮   ┌───┐   ╭───────────╮
+│ fetch_add ├─┐ │ 0 │ ┌─┤ fetch_add │
+╰───────────╯ │ └───┘ │ ╰───────────╯
+              └─┬───┐ │
+                │ 1 │ │
+                └───┘ │
+                ┌───┬─┘
+                │ 2 │
+                └───┘
+```
+
+There is one problem with this code however, and that is that if `get_id()` is
+called over 18 446 744 073 709 551 615 times, the counter will overflow and it
+will start generating duplicate IDs. Of course, this won’t feasibly happen, but
+it can be problematic if you need to _prove_ that it can’t happen (e.g. for
+safety purposes) or you’re using a smaller integer type like `u32`.
+
+So we’re going to modify this function so that instead of returning a plain
+`u64` it returns an `Option<u64>`, where `None` is used to indicate that an
+overflow occurred and no more IDs could be generated. Additionally, it’s not
+enough to just return `None` once, because if there are multiple threads
+involved they will not see that result if it just occurs on a single thread —
+instead, it needs to continue to return `None` _until the end of time_ (or,
+well, this execution of the program).
+
+That means we have to do away with `fetch_add`, because `fetch_add` will always
+overflow and there’s no `checked_fetch_add` equivalent. We’ll return to our racy
+algorithm for a minute, this time thinking more about what went wrong. The steps
+look something like this:
+
+1. Load a value of the atomic
+1. Perform the checked add, propagating `None`
+1. Store in the new value of the atomic
+
+The problem here is that the store does not necessarily occur directly after the
+load in the atomic’s modification order, and that leads to the races. What we
+need is some way to say, “add this new value to the modification order, but
+_only if_ it occurs directly after the value we loaded”. And luckily for us,
+there exists a function that does exactly\* this: `compare_exchange`.
+
+`compare_exchange` is a bit like a store, but instead of unconditionally storing
+the value, it will first check the previous value in the modification order to
+see whether it is what we expect, and if not it will simply tell us that and not
+make any changes. It is an RMW operation, so all of this happens fully
+atomically — there is no chance for a race condition.
+
+> \* It’s not quite the same, because `compare_exchange` can suffer from ABA
+> problems in which it will see a later value in the modification order that
+> just happened to be same and succeed. However, in this code values can never
+> be reused so we don’t have to worry about that.
```

Comment:
This is the first time "ABA problem" is mentioned and it might be helpful to put some concrete values into a one-sentence example. Like "for instance, the modification order could contain the sub-sequence (1, 2, 1), and while the `compare_exchange` was intended to be performed on the first 1, it would also succeed on the second."